docs/proposals/001-in-place-upgrades.md

```diff
@@ -0,0 +1,257 @@
+* `k8sd.io/in-place-upgrade-failure` annotation on the `Machine` would be added/updated with the failure message
+* `k8sd.io/in-place-upgrade-status` annotation on the `Machine` would be added/updated with `failed`
+
+The reconciler should ignore the upgrade if `k8sd.io/in-place-upgrade-status` is already set to `in-progress` on the machine. 
+
+#### Changes for Rolling Upgrades and Creating New Machines
+In case of a rolling upgrade or when creating new machines the `CK8sConfigReconciler` should check for the `k8sd.io/in-place-upgrade-current` annotation both on the `Machine` and on the owner `Cluster` object.
+
+The value of one of these annotations should be used instead of the `version` field while generating a cloud-init script for a machine. The precedence of version fields are:
+1. Annotation on the `Machine`
+2. Annotation on the `Cluster`
+3. The `version` field
+
+Which means the value from the annotation on the `Machine` would be used first if found.
+
+Using an annotation value requires changing the `install.sh` file to perform the relevant snap operation based on the option.
+* `snap install k8s --classic --channel <channel>` for `Channel`
+* `snap install k8s --classic --revision <revision>` for `Revision`
+* `snap install <path-to-snap> --classic --dangerous --name k8s` for `LocalPath`
+
+When a rolling upgrade is triggered the `LocalPath` option requires the newly created machine to contain the local `*.snap` file. This usually means the machine image used by the infrastructure provider should be updated to contain this image. This file could possibly be sideloaded in the cloud-init script before installation.
+
+This operation should not be performed if `install.sh` script is overridden by the user in the manifests.
+
+This would prevent adding nodes with an outdated version and possibly breaking the cluster due to a version mismatch.
+
+## ControlPlane Provider Changes
+<!--
+This section MUST mention any changes to the controlplane provider.
+-->
+A cluster controller called `ClusterReconciler` is added which would perform the one-by-one in-place upgrade of the entire workload cluster. 
+
+The controller would propagate the `k8sd.io/in-place-upgrade-to` annotation on the `Cluster` object by adding this annotation one-by-one to all the machines that is owned by this cluster. 
+
+A Kubernetes API call listing the objects of type `Machine` and filtering with `ownerRef` would produce the list of machines owned by the cluster. The controller then would iterate over this list, annotating machines and waiting for the operation to complete on each iteration.
+
+The reconciler should ignore a machine if `k8sd.io/in-place-upgrade-status` is already set to `in-progress`. 
+
+Once upgrades of the underlying machines are finished:
+* `k8sd.io/in-place-upgrade-to` annotation on the `Cluster` would be removed
+* `k8sd.io/in-place-upgrade-current` annotation on the `Cluster` would be added/updated with the used `{upgrade-option}`.
```

Comment:
This is a good approach. I'd recommend moving this as "Out Of Scope" to start with, until we have some confidence on the feature itself. It would be easy to lose cycles iterating on this orchestration layer, rather than building and testing out the in place upgrades.

"Out of scope" means: keep the text, _but_ do not commit to an implementation decision just yet.